Source/DataStructureAndEncodingDefinition/fixed_istream_buffer.h

```diff
@@ -0,0 +1,84 @@
+#include <streambuf>
+#include <cstring>
+#include <vector>
+#include <istream>
+
+/**
+ gdcmReader does seeking when decoding the headers so we need to buffer the input stream
+ can seek back at least half the buffer size, fails when seeking outside of the buffer
+ may make a blocking read on construction
+ based on https://stackoverflow.com/questions/44711735/someway-to-buffer-or-wrap-cin-so-i-can-use-tellg-seekg
+*/
+class fixed_istream_buffer : public std::streambuf {
+    std::istream& stream;
+    const unsigned size;
+    const unsigned half;
+    std::vector<char> buffer;
+    std::streamoff base;
+
+public:
+    fixed_istream_buffer(std::istream& stream, unsigned size)
+        : stream(stream), size(size), half(size / 2), buffer(size), base() {
+        std::streamsize read = stream.read(buffer.data(), size).gcount();
+        setg(buffer.data(), buffer.data(), buffer.data() + read);
+    }
+
+    int underflow() override {
+        if (gptr() == buffer.data() + size) {
+            char* half_buffer = eback() + half;
+            std::memmove(eback(), half_buffer, half);
+            std::streamsize read = stream.read(half_buffer, half).gcount();
+            setg(eback(), half_buffer, half_buffer + read);
+            base += half;
```

Comment:
I'm not seeing where `base` is ever reset? It seems to be ever-increasing?

Reply:
base corresponds to the position of eback (the first element of the buffer) relative to the input stream